File: app/src/main/java/ca/appvelopers/mcgillmobile/util/thread/UserDownloader.java
/*
 * Copyright 2014-2016 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.appvelopers.mcgillmobile.util.thread;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.LocalBroadcastManager;

import com.guerinet.utils.Utils;

import java.io.IOException;
import java.util.List;

import javax.inject.Inject;

import ca.appvelopers.mcgillmobile.App;
import ca.appvelopers.mcgillmobile.R;
import ca.appvelopers.mcgillmobile.model.Course;
import ca.appvelopers.mcgillmobile.model.Semester;
import ca.appvelopers.mcgillmobile.model.Term;
import ca.appvelopers.mcgillmobile.model.Transcript;
import ca.appvelopers.mcgillmobile.model.exception.MinervaException;
import ca.appvelopers.mcgillmobile.model.retrofit.McGillService;
import ca.appvelopers.mcgillmobile.util.Constants;
import ca.appvelopers.mcgillmobile.util.manager.ScheduleManager;
import ca.appvelopers.mcgillmobile.util.manager.TranscriptManager;
import timber.log.Timber;

/**
 * Downloads the user's information
 * @author Julien Guerinet
 * @since 2.2.0
 */
public abstract class UserDownloader extends Thread {
    /**
     * App context
     */
    @Inject
    protected Context context;
    /**
     * {@link McGillService} instance
     */
    @Inject
    protected McGillService mcGillService;
    /**
     * {@link TranscriptManager} instance
     */
    @Inject
    protected TranscriptManager transcriptManager;
    /**
     * {@link ScheduleManager} instance
     */
    @Inject
    protected ScheduleManager scheduleManager;
    /**
     * True if everything should be downloaded, false otherwise (defaults to false)
     */
    protected boolean downloadEverything = false;
    /**
     * {@link IOException} that potentially occurred during the downloading of info
     */
    private IOException exception;

    /**
     * Default Constructor
     *
     * @param context App context
     */
    public UserDownloader(Context context) {
        App.component(context).inject(this);
    }

    @Override
    public void run() {
        synchronized (this) {
            //If we're not connected to the internet, don't continue
            if (!Utils.isConnected(context)) {
                return;
            }

            //Transcript
            if (downloadEverything) {
                update(context.getString(R.string.downloading_transcript));
            }

            //Set the old transcript instance to use for the semesters if ever we don't get to
            //  download the new one
            Transcript transcript = transcriptManager.get();
            try {
                transcript = mcGillService.transcript().execute().body();
                transcriptManager.set(transcript);
            } catch(IOException e) {
                if (e instanceof MinervaException) {
                    LocalBroadcastManager.getInstance(context)
                            .sendBroadcast(new Intent(Constants.BROADCAST_MINERVA));
                } else {
                    Timber.e(e, "Transcript Exception");
                }
                exception = e;
            }

            //The current term
            Term currentTerm = Term.currentTerm();

            //Go through the semesters
            for (Semester semester: transcript.getSemesters()) {
                //Get the term of this semester
                Term term = semester.getTerm();

                //If we are not downloading everything, only download it if it's the
                //  current or future term
                if (downloadEverything || term.equals(currentTerm) ||
                        term.isAfter(currentTerm)) {
                    if (downloadEverything) {
                        update(context.getString(R.string.downloading_semester,
                                term.getString(context)));
                    }

                    //Download the schedule
                    try {
                        List<Course> courses = mcGillService.schedule(term).execute().body();
                        scheduleManager.set(courses, term);
                    } catch (IOException e) {
                        if (e instanceof MinervaException) {
                            LocalBroadcastManager.getInstance(context)
                                    .sendBroadcast(new Intent(Constants.BROADCAST_MINERVA));
                        } else {
                            Timber.e(e, "Term Exception: %s", term.getId());
                        }
                        exception = e;
                    }
                }
            }

            //Ebill
            if (downloadEverything) {
                update(context.getString(R.string.downloading_ebill));
            }

            //Download the eBill and user info
            try {
                App.setEbill(mcGillService.ebill().execute().body());
            } catch(IOException e) {
                if (e instanceof MinervaException) {
                    LocalBroadcastManager.getInstance(context)
                            .sendBroadcast(new Intent(Constants.BROADCAST_MINERVA));
                } else {
                    Timber.e(e, "Ebill Exception");
                }
                exception = e;
            }
            notify();
        }
    }

    /**
     * Executes the thread synchronously
     *
     * @throws IOException
     */
    public void execute() throws IOException {
        //If we are downloading this synchronously, then we're downloading everything
        downloadEverything = true;

        //Start the download
        start();

        //Wait until the thread has been fully executed
        synchronized(this) {
            try {
                wait();
            } catch(InterruptedException ignored){}
        }

        //If there's an exception, throw it
        if (exception != null) {
            throw exception;
        }
    }

    public abstract void update(String section);
}

File: app/src/main/java/util/thread/UserDownloader.kt
/*
 * Copyright 2014-2022 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.util.thread

import android.content.Context
import android.content.Intent
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.model.Term
import com.guerinet.mymartlet.model.exception.MinervaException
import com.guerinet.mymartlet.util.Constants
import com.guerinet.mymartlet.util.retrofit.McGillService
import com.guerinet.mymartlet.util.room.daos.CourseDao
import com.guerinet.mymartlet.util.room.daos.SemesterDao
import com.guerinet.mymartlet.util.room.daos.StatementDao
import com.guerinet.mymartlet.util.room.daos.TranscriptCourseDao
import com.guerinet.mymartlet.util.room.daos.TranscriptDao
import com.guerinet.suitcase.coroutines.bgDispatcher
import com.guerinet.suitcase.coroutines.uiDispatcher
import com.guerinet.suitcase.util.extensions.isConnected
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.locks.ReentrantLock
import javax.net.ssl.SSLException
import kotlin.concurrent.withLock

/**
 * TODO
 * Downloads the user's information
 * @author Julien Guerinet
 * @since 2.2.0
 */
abstract class UserDownloader(val context: Context) : Thread(), KoinComponent {

    private val mcGillService by inject<McGillService>()

    private val courseDao by inject<CourseDao>()

    private val semesterDao by inject<SemesterDao>()

    private val statementDao by inject<StatementDao>()

    private val transcriptDao by inject<TranscriptDao>()

    private val transcriptCourseDao by inject<TranscriptCourseDao>()

    private val lock = ReentrantLock()

    private val condition = lock.newCondition()

    /**
     * True if everything should be downloaded, false otherwise (defaults to false)
     */
    protected var downloadEverything = false
    /**
     * [IOException] that potentially occurred during the downloading of info
     */
    private var exception: IOException? = null

    override fun run() {
        lock.withLock {
            //If we're not connected to the internet, don't continue
            if (!context.isConnected) {
                return
            }

            //Transcript
            if (downloadEverything) {
                update(context.getString(R.string.downloading_transcript))
            }

            try {
                val transcriptResponse = mcGillService.oldTranscript().execute().body()
                if (transcriptResponse != null) {
                    transcriptDao.insert(transcriptResponse.transcript)
                    transcriptCourseDao.update(transcriptResponse.courses)
                    semesterDao.update(transcriptResponse.semesters)
                }
            } catch (e: IOException) {
                handleException(e, "Transcript")
            }

            val currentTerm = Term.currentTerm()

            // Go through the semesters
            val semesters = semesterDao.getSemesters()
            for (semester in semesters) {
                //Get the currentTerm of this semester
                val term = semester.term

                //If we are not downloading everything, only download it if it's the
                //  current or future currentTerm
                if (downloadEverything || term == currentTerm || term.isAfter(currentTerm)) {
                    if (downloadEverything) {
                        update(
                            context.getString(
                                R.string.downloading_semester,
                                term.getString(context)
                            )
                        )
                    }

                    //Download the schedule
                    try {
                        val courses = mcGillService.schedule(term).execute().body()
                        if (courses != null) {
                            courseDao.update(courses, term)
                        }
                    } catch (e: IOException) {
                        handleException(e, term.id)
                    }
                }
            }

            // Ebill
            if (downloadEverything) {
                update(context.getString(R.string.downloading_ebill))
            }

            GlobalScope.launch(uiDispatcher) {
                try {
                    val response = mcGillService.ebill().await()
                    withContext(bgDispatcher) {
                        statementDao.update(response)
                    }
                } catch (e: IOException) {
                    handleException(e, "Ebill")
                }
            }
            condition.signal()
        }
    }

    /**
     * Handles any exception while downloading the user info
     *
     * @param e Exception instance
     * @param section Section that the exception happened in
     */
    private fun handleException(e: IOException, section: String) {
        if (e is MinervaException) {
            androidx.localbroadcastmanager.content.LocalBroadcastManager.getInstance(context)
                .sendBroadcast(Intent(Constants.BROADCAST_MINERVA))
        } else if (e !is SSLException) {
            // Don't log SSLExceptions
            Timber.e(Exception("Exception: $section", e))
        }
        exception = e
    }

    /**
     * Executes the thread synchronously
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    fun execute() {
        //If we are downloading this synchronously, then we're downloading everything
        downloadEverything = true

        //Start the download
        start()

        //Wait until the thread has been fully executed
        lock.withLock {
            condition.await()
        }

        //If there's an exception, throw it
        exception?.apply { throw this }
    }

    abstract fun update(section: String)
}
